### PR TITLE
#792 lock not allow refund

### PR DESCRIPTION
**Pull request by vietdangquoc**
offer đang có order OPEN PENDING CONFIRM EXPIRED thì không được cancel add status EXPIRED
CANCELED = Manual cancellation by user
EXPIRED = Automatic expiration
FAILED = Blockchain transaction failed during order confirmation

# Ticket
#792

# General
- [ ] No unnecessary files, debug logs, or commented code included.
- [ ] PR has at least 1 peer approval before senior review.
- [ ] I fully understand the code I submitted.
- [ ] No hardcoded configuration or secrets.
- [ ] **Environment Variables** changes? (Leave unchecked if N/A)

# Design & Solution Clarity
- [ ] A design diagram (sequence / flow) is attached or linked.
- [ ] The implementation matches the design.
- [ ] All major flows and decision points are represented in the diagram.
- [ ] Error handling and edge cases are clearly defined in the design.

Design link (diagram / doc):
`<paste link here>`

# Testing
- [ ] Unit tests cover main logic and important edge cases.

> Note: Complex logic must not rely on manual testing only.

# UI evidence (For frontend only)
`<paste screenshots here>`

# Peer Review Checklist
- [ ] I understand the overall solution and flow.
- [ ] The code follows the provided design diagram.
- [ ] No obvious logic or security issues found.
- [ ] Tests are sufficient for the scope of this change.

### CODE REVIEW

**Review by phongvu2812 — APPROVED** [review]

File: dong-service/scheduler/cancel_expired_orders_job.go [hunk 1]
```diff
@@ -79,7 +79,15 @@ func (j *CancelExpiredOrdersJob) Run(ctx context.Context) error {
...
-		if orderInfo.Status != constants.TradingPending {
+		if orderInfo.PreviousStatus == constants.TradingPending {
+			logger.Info().
+				Int64("order_id", orderInfo.OrderID).
+				Str("previous_status", orderInfo.PreviousStatus).
+				Msg("Skipping refund: order reached PENDING status (payment was confirmed)")
+			continue
+		}
+
+		if orderInfo.Status != constants.TradingPending && orderInfo.Status != constants.TradingWaiting {
```
Comment (QTR33) [inline dong-service/scheduler/cancel_expired_orders_job.go]: trạng thái open

**Review by phongvu2812 — APPROVED** [review]

File: dong-service/services/p2p_offer_service.go [hunk 2]
```diff
@@ -322,11 +322,20 @@ func (s *OfferService) CancelOffer(ctx context.Context, offerId int64, offer *mo
 		}
 	}()
 
-	hasActive, err := s.orderRepo.HasActiveOrders(ctx, offerId, tx)
+	// Block cancellation if any order is in dispute (expired while buyer had already confirmed payment).
+	hasDisputed, err := s.orderRepo.HasDisputedOrders(ctx, offerId)
```
Comment (hoangdohuy) [inline dong-service/services/p2p_offer_service.go]: phải dùng trong cùng 1 transaction của database

File: dong-service/repository/p2p_order_repository.go [hunk 3]
```diff
@@ -63,9 +76,9 @@ func (r *OrderRepository) CountActiveOrdersByUser(ctx context.Context, buyerUser
 func (r *OrderRepository) UpdateOrderStatus(ctx context.Context, orderID int64, status string, tx *sql.Tx) error {
 	query := fmt.Sprintf(`
 		UPDATE %s.p2p_orders
-		SET status = $1, updated_at = NOW()
+		SET status = $1, previous_status = (SELECT status FROM %s.p2p_orders WHERE order_id = $2), updated_at = NOW()
```
Comment (hoangdohuy) [inline dong-service/repository/p2p_order_repository.go]: xem lại hình như có thể update gọn hơn

File: dong-service/repository/p2p_order_repository.go [hunk 4]
```diff
@@ -75,9 +88,9 @@ func (r *OrderRepository) UpdateOrderStatusWithTxHash(ctx context.Context, order
 	if txHash != nil {
 		query := fmt.Sprintf(`
 			UPDATE %s.p2p_orders
-			SET status = $1, transaction_hash = $2, updated_at = NOW()
+			SET status = $1, transaction_hash = $2, previous_status = (SELECT status FROM %s.p2p_orders WHERE order_id = $3), updated_at = NOW()
```
Comment (hoangdohuy) [inline dong-service/repository/p2p_order_repository.go]: cài này cũng thế

File: dong-service/migrations/031_add_previous_status_and_dispute_columns.sql [hunk 5]
```diff
@@ -0,0 +1,14 @@
+-- Add previous_status column to p2p_orders table to track status transitions
```
Comment (hoangdohuy) [inline dong-service/migrations/031_add_previous_status_and_dispute_columns.sql:1]: xem lại version ở dev

File: dong-service/migrations/030_add_expired_status_to_orders.sql [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+BEGIN
+    IF NOT EXISTS (
+        SELECT 1
+        FROM pg_enum
+        WHERE enumlabel = 'EXPIRED'
+          AND enumtypid = (
+              SELECT oid FROM pg_type WHERE typname = 'order_status'
+          )
+    ) THEN
+        ALTER TYPE order_status ADD VALUE 'EXPIRED';
```
Comment (hoangdohuy) [inline dong-service/migrations/030_add_expired_status_to_orders.sql:12]: xem lại version ở dev

File: dong-service/services/p2p_offer_service.go [hunk 7]
```diff
@@ -322,11 +322,20 @@ func (s *OfferService) CancelOffer(ctx context.Context, offerId int64, offer *mo
...
+	// Block cancellation if any order is in dispute (expired while buyer had already confirmed payment).
+	hasDisputed, err := s.orderRepo.HasDisputedOrders(ctx, offerId, tx)
+	if err != nil {
+		return fmt.Errorf("failed to check disputed orders: %w", err)
+	}
+	if hasDisputed {
+		return fmt.Errorf("cannot cancel offer: one or more orders are in dispute")
+	}
+
+	orderCount, err := s.orderRepo.CountOrdersByOffer(ctx, offerId)
```
Comment (hoangdohuy) [inline dong-service/services/p2p_offer_service.go]: wrap vào transaction

**Review by hoangdohuy — APPROVED** [review]
LGTM
